### PR TITLE
[Tune] Add ZOOpt search algorithm

### PR DESCRIPTION
**Pull request by zewenli98**
## Why are these changes needed?
[ZOOpt](https://github.com/eyounx/ZOOpt/tree/master/zoopt) is a python package for Zeroth-Order Optimization (a.k.a. derivative-free optimization/black-box optimization).

This package allows users to mix continuous dimensions and discrete dimensions.

I implemented ASRacos search algorithm that could be easily harnessed to optimize tasks as below:

```python
from ray.tune.suggest.zoopt import ZOOptSearch
from zoopt import ValueType
from ray.tune import run

def easy_objective(config, reporter):
    import time
    time.sleep(0.2)
    for i in range(config["iterations"]):
        reporter(
            timesteps_total=i,
            mean_loss=(config["height"] - 14)**2 - abs(config["width"] - 3))
        time.sleep(0.02)

# This dict could mix continuous dimensions and discrete dimensions, for example:
dim_dict = {
    # for continuous dimensions: (continuous, search_range, precision)
    "height": (ValueType.CONTINUOUS, [-10, 10], 1e-2), 
    # for discrete dimensions: (discrete, search_range, has_order)
    "width": (ValueType.DISCRETE, [-10, 10], False)
}

config = {
        "num_samples": 200,
        "config": {
            "iterations": 1,  # evaluation times
        },
        "stop": {
            "timesteps_total": 1  # cumstom stop rules
        }
    }

zoopt_search = ZOOptSearch(
        algo="Asracos",  # only support Asracos currently
        budget=config["num_samples"],
        dim_dict=dim_dict,
        max_concurrent=16,
        metric="mean_loss",
        mode="min"
        )

run(easy_objective,
        search_alg=zoopt_search,
        **config)
```

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/tune/suggest/__init__.py [hunk 1]
```diff
@@ -34,3 +34,8 @@ def SkOptSearch(*args, **kwargs):
 def SigOptSearch(*args, **kwargs):
     raise DeprecationWarning("""This class has been moved. Please import via
         `from ray.tune.suggest.sigopt import SigOptSearch`""")
+
+
+def ZOOptSearch(*args, **kwargs):
+    raise DeprecationWarning("""This class has been moved. Please import via
+        `from ray.tune.suggest.zoopt import ZOOptSearch`""")
```
Comment (richardliaw) [inline python/ray/tune/suggest/__init__.py:41]: no need for this

File: python/ray/tune/suggest/zoopt.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
+import copy
+import logging
+import pickle
+
+try:  # Python 3 only -- needed for lint test.
+    from zoopt import Dimension2, Parameter
+    from zoopt.algos.opt_algorithms.racos.sracos import SRacosTune
+except ImportError:
+    zoopt = None
```
Comment (richardliaw) [inline python/ray/tune/suggest/zoopt.py:9]: We already only support python3, so no need for this catch

**Review by richardliaw — CHANGES_REQUESTED** [review]
Thank you so much for the contribution! Would it be possible to change the same files as https://github.com/ray-project/ray/pull/5955 to make sure:

1. ZOOpt is documented
2. ZOOpt is tested

Thanks!

**Comment by Servon-Lee** [comment]
I followed #5955 but got the problem:

``` bash
==================================== ERRORS ====================================
_____________ ERROR collecting ray/tune/tests/test_tune_restore.py _____________
ImportError while importing test module '/ray/python/ray/tune/tests/test_tune_restore.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
python/ray/tune/tests/test_tune_restore.py:20: in <module>
    from ray.tune.suggest.zoopt_search import ZOOptSearch
/opt/conda/lib/python3.6/site-packages/ray/tune/suggest/zoopt_search.py:4: in <module>
    from zoopt import Dimension2, Parameter
E   ModuleNotFoundError: No module named 'zoopt'
```
How can I fix it? @richardliaw

**Comment by richardliaw** [comment]
Ah due to a recent CI change, we moved our requirement file - 
Can you add `zoopt` to:
```ray/docker/tune_test/requirements.txt```

**Comment by Servon-Lee** [comment]
> Ah due to a recent CI change, we moved our requirement file -
> Can you add `zoopt` to:
> `ray/docker/tune_test/requirements.txt`

Thanks man, it works. But a new problem occurs. I feel like these failures have nothing to do with zoopt. Can you take a look? Thanks a ton!

``` bash
=================================== FAILURES ===================================
________________________ test_remove_node_before_result ________________________

start_connected_emptyhead_cluster = <ray.cluster_utils.Cluster object at 0x7f20c0580048>

    def test_remove_node_before_result(start_connected_emptyhead_cluster):
        """Tune continues when node is removed before trial returns."""
        cluster = start_connected_emptyhead_cluster
        node = cluster.add_node(num_cpus=1)
        cluster.wait_for_nodes()
    
        runner = TrialRunner(BasicVariantGenerator())
        kwargs = {
            "stopping_criterion": {
                "training_iteration": 3
            },
            "checkpoint_freq": 2,
            "max_failures": 2
        }
        trial = Trial("__fake", **kwargs)
        runner.add_trial(trial)
    
        runner.step()  # Start trial
        assert trial.status == Trial.RUNNING
        cluster.remove_node(node)
        cluster.add_node(num_cpus=1)
        cluster.wait_for_nodes()
        assert ray.cluster_resources()["CPU"] == 1
    
        # Process result (x2), process save, process result.
        for _ in range(4):
            runner.step()
>       assert trial.status == Trial.TERMINATED
E       AssertionError: assert 'RUNNING' == 'TERMINATED'
E         - TERMINATED
E         + RUNNING

python/ray/tune/tests/test_cluster.py:159: AssertionError
---------------------------- Captured stderr setup -----------------------------
2020-04-15 05:00:27,353	INFO resource_spec.py:212 -- Starting Ray with 5.71 GiB memory available for workers and up to 0.15 GiB for objects. You can adjust these settings with ray.init(memory=<bytes>, object_store_memory=<bytes>).
2020-04-15 05:00:27,657	INFO services.py:1150 -- View the Ray dashboard at localhost:8265
2020-04-15 05:00:27,763	WARNING worker.py:785 -- When connecting to an existing cluster, _internal_config must match the cluster's _internal_config.
2020-04-15 05:00:27,817	INFO cluster_utils.py:58 -- {'node_ip_address': '172.17.0.2', 'redis_address': '172.17.0.2:17636', 'object_store_address': '/tmp/ray/session_2020-04-15_05-00-27_352243_1/sockets/plasma_store', 'raylet_socket_name': '/tmp/ray/session_2020-04-15_05-00-27_352243_1/sockets/raylet', 'webui_url': 'localhost:8265', 'session_dir': '/tmp/ray/session_2020-04-15_05-00-27_352243_1'}
----------------------------- Captured stdout call -----------------------------
(pid=raylet) E0415 05:00:28.256134   365 process.cc:274] Failed to wait for process 377 with error system:10: No child processes
(pid=395) unable to import 'smart_open.gcs', disabling that module
Result for __fake_0649f1c8:
  date: 2020-04-15_05-00-32
  done: false
  episode_len_mean: 10
  episode_reward_mean: 10
  experiment_id: 74ed85b11cd04136999fb71727329f0d
  hostname: 1bbe4032e79b
  info: {}
  iterations_since_restore: 1
  node_ip: 172.17.0.2
  perf:
    cpu_util_percent: 64.1
    ram_util_percent: 24.5
  pid: 395
  time_since_restore: 5.245208740234375e-06
  time_this_iter_s: 5.245208740234375e-06
  time_total_s: 5.245208740234375e-06
  timestamp: 1586926832
  timesteps_since_restore: 10
  timesteps_this_iter: 10
  timesteps_total: 10
  training_iteration: 1
  trial_id: 0649f1c8
  
----------------------------- Captured stderr call -----------------------------
2020-04-15 05:00:28,030	INFO resource_spec.py:212 -- Starting Ray with 6.2 GiB memory available for workers and up to 0.15 GiB for objects. You can adjust these settings with ray.init(memory=<bytes>, object_store_memory=<bytes>).
E0415 05:00:28.254688   354 task_manager.cc:306] Task failed: IOError: cancelling all pending tasks of dead actor: Type=ACTOR_TASK, Language=PYTHON, function_descriptor={type=PythonFunctionDescriptor, module_name=ray.rllib.agents.mock, class_name=_MockTrainer, function_name=train, function_hash=}, task_id=8603ac10b8c093b315c675b20100, job_id=0100, num_args=0, num_returns=2, actor_task_spec={actor_id=15c675b20100, actor_caller_id=ffffffffffffffffffffffff0100, actor_counter=0}
2020-04-15 05:00:28,268	INFO resource_spec.py:212 -- Starting Ray with 6.2 GiB memory available for workers and up to 0.15 GiB for objects. You can adjust these settings with ray.init(memory=<bytes>, object_store_memory=<bytes>).
2020-04-15 05:00:28,411	ERROR trial_runner.py:521 -- Trial __fake_0649f1c8: Error processing event.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/ray/tune/trial_runner.py", line 467, in _process_trial
    result = self.trial_executor.fetch_result(trial)
  File "/opt/conda/lib/python3.6/site-packages/ray/tune/ray_trial_executor.py", line 431, in fetch_result
    result = ray.get(trial_future[0], DEFAULT_GET_TIMEOUT)
  File "/opt/conda/lib/python3.6/site-packages/ray/worker.py", line 1490, in get
    raise value
+ [[ 1 -lt 3 ]]
+ (( n++ ))
ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
E0415 05:00:28.412922     1 task_manager.cc:306] Task failed: IOError: sent task to dead actor: Type=ACTOR_TASK, Language=PYTHON, function_descriptor={type=PythonFunctionDescriptor, module_name=ray.rllib.agents.mock, class_name=_MockTrainer, function_name=stop, function_hash=}, task_id=276fa5c994b5ce4315c675b20100, job_id=0100, num_args=0, num_returns=2, actor_task_spec={actor_id=15c675b20100, actor_caller_id=ffffffffffffffffffffffff0100, actor_counter=1}
2020-04-15 05:00:28,413	INFO trial_runner.py:636 -- Trial __fake_0649f1c8: Attempting to restore trial state from last checkpoint.
2020-04-15 05:00:29,860	WARNING worker.py:1065 -- The dashboard on node 1bbe4032e79b failed with the following error:
+ echo 'Command failed. Attempt 2/3:'
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/ray/dashboard/dashboard.py", line 1183, in <module>
    metrics_export_address=metrics_export_address)
  File "/opt/conda/lib/python3.6/site-packages/ray/dashboard/dashboard.py", line 513, in __init__
    build_dir = setup_static_dir(self.app)
  File "/opt/conda/lib/python3.6/site-packages/ray/dashboard/dashboard.py", line 418, in setup_static_dir
    "&& npm run build)", build_dir)
+ true
+ bash /home/runner/work/ray/ray/ci/jenkins_tests/run_tune_tests.sh 8G 4G
FileNotFoundError: [Errno 2] Dashboard build directory not found. If installing from source, please follow the additional steps required to build the dashboard(cd python/ray/dashboard/client && npm ci && npm run build): '/opt/conda/lib/python3.6/site-packages/ray/dashboard/client/build'
```

**Comment by Servon-Lee** [comment]
@richardliaw

**Review by richardliaw — APPROVED** [review]
Looks good to me - test failures look unrelated.

File: doc/source/tune-searchalg.rst [hunk 3]
```diff
@@ -286,6 +286,59 @@ Take a look at `an example here <https://github.com/ray-project/ray/blob/master/
     :show-inheritance:
     :noindex:
 
+ZOOpt Search
+------------
+
+The ``ZOOptSearch`` is a list of derivative-free optimization algorithms that are backed by `ZOOpt <https://github.com/polixir/ZOOpt>`__ package to perform sequential model-based hyperparameter optimization. Currently, Asynchronous Sequential RAndomized COordinate Shrinking (ASRacos) algorithm is implemented in Tune. Note that this class does not extend ``ray.tune.suggest.BasicVariantGenerator``, so you will not be able to use Tune’s default variant generation/search space declaration when using ZOOptSearch.
```
Comment (richardliaw) [inline doc/source/tune-searchalg.rst:292]:
```suggestion
The ``ZOOptSearch`` is a SearchAlgorithm for derivative-free optimization. It is backed by the `ZOOpt <https://github.com/polixir/ZOOpt>`__ package. Currently, Asynchronous Sequential RAndomized COordinate Shrinking (ASRacos) algorithm is implemented in Tune. Note that this class does not extend ``ray.tune.suggest.BasicVariantGenerator``, so you will not be able to use Tune’s default variant generation/search space declaration when using ZOOptSearch.
```

**Pull request merged**